File: MidtermLabWork3/java/MP17.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * MP17 - Find the longest text entry in dataset.
 * Student: CELADEZ, JED CEDRIC G.
 */
public class MP17 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter dataset file path: ");
        String filePath = scanner.nextLine().trim();

        List<String> headers = new ArrayList<>();
        List<List<String>> rows = new ArrayList<>();

        try {
            loadDataset(filePath, headers, rows);
            runMP17(headers, rows);
        } catch (IOException e) {
            System.out.println("Error: Unable to read file. " + e.getMessage());
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }

    private static void runMP17(List<String> headers, List<List<String>> rows) {
        String longestEntry = "";
        int longestRow = -1;
        int longestCol = -1;

        for (int r = 0; r < rows.size(); r++) {
            List<String> row = rows.get(r);
            for (int c = 0; c < row.size(); c++) {
                String value = row.get(c).trim();
                if (value.length() > longestEntry.length()) {
                    longestEntry = value;
                    longestRow = r;
                    longestCol = c;
                }
            }
        }

        System.out.println("\n--- Longest Text Entry Found ---");
        if (longestRow == -1) {
            System.out.println("No valid data found.");
            return;
        }

        String columnName = (longestCol >= 0 && longestCol < headers.size())
                ? headers.get(longestCol)
                : "Column " + (longestCol + 1);

        System.out.println("Text: " + longestEntry);
        System.out.println("Length: " + longestEntry.length() + " characters");
        System.out.println("Location: Row " + (longestRow + 1) + ", Column: " + columnName);
    }

    private static void loadDataset(String filePath, List<String> headers, List<List<String>> rows)
            throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            boolean headerFound = false;

            while ((line = reader.readLine()) != null) {
                if (line.trim().isEmpty()) {
                    continue;
                }

                List<String> cells = parseCsvLine(line);

                if (!headerFound) {
                    boolean isHeader = false;
                    for (String cell : cells) {
                        if ("candidate".equalsIgnoreCase(cell.trim())) {
                            isHeader = true;
                            break;
                        }
                    }

                    if (isHeader) {
                        headers.addAll(cells);
                        headerFound = true;
                    }
                    continue;
                }

                rows.add(cells);
            }
        }

        if (headers.isEmpty()) {
            throw new IllegalArgumentException("Dataset header row not found.");
        }
    }

    private static List<String> parseCsvLine(String line) {
        List<String> cells = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    current.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                cells.add(current.toString());
                current.setLength(0);
            } else {
                current.append(ch);
            }
        }

        cells.add(current.toString());
        return cells;
    }
}

File: MidtermLabWork3/javascript/MP17.js
const fs = require('fs');
const readline = require('readline');

/**
 * MP17 – Find the longest text entry in dataset.
 * Student: CELADEZ, JED CEDRIC G.
 */

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('Enter dataset file path: ', (filePath) => {
    if (!fs.existsSync(filePath)) {
        console.error('Error: File not found.');
        rl.close();
        return;
    }

    try {
        const fileContent = fs.readFileSync(filePath, 'utf8');
        const lines = fileContent.split(/\r?\n/);
        
        let longestEntry = "";
        let longestRow = -1;
        let longestHeader = "";
        let headers = [];
        let headerFound = false;
        let currentRow = 0;

        for (let line of lines) {
            // Skip empty rows and rows that don't contain the anchor "Candidate"
            if (line.trim() === "" || (!headerFound && !line.includes("Candidate"))) {
                continue;
            }

            const cells = parseCsvLine(line);

            if (!headerFound) {
                headers = cells;
                headerFound = true;
                continue;
            }

            currentRow++;
            cells.forEach((cell, index) => {
                const trimmedCell = cell.trim();
                if (trimmedCell.length > longestEntry.length) {
                    longestEntry = trimmedCell;
                    longestRow = currentRow;
                    longestHeader = headers[index] || `Column ${index}`;
                }
            });
        }

        if (longestRow !== -1) {
            console.log("\n--- Longest Text Entry Found ---");
            console.log("Text: ", longestEntry);
            console.log("Length: ", longestEntry.length, " characters");
            console.log("Location: Row ", longestRow, ", Column: ", longestHeader);
        } else {
            console.log("No valid data found.");
        }

    } catch (err) {
        console.error('Error processing file:', err.message);
    } finally {
        rl.close();
    }
});

/**
 * Parses a CSV line handling quoted strings and commas.
 */
function parseCsvLine(line) {
    const cells = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && line[i + 1] === '"') {
                current += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === ',' && !inQuotes) {
            cells.push(current);
            current = '';
        } else {
            current += ch;
        }
    }

    cells.push(current);
    return cells;
}
